src/Services/SessionState/SessionStateCollection.cs

```diff
@@ -0,0 +1,215 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters.SessionState.Serialization;
+
+internal class SessionStateCollection : ISessionState
+{
+    private readonly Dictionary<string, ItemHolder> _items;
+
+    public SessionStateCollection(ISessionKeySerializer serializer)
+    {
+        Serializer = serializer;
+        _items = [];
+    }
+
+    protected SessionStateCollection(SessionStateCollection other)
+    {
+        _items = other._items;
+
+        Serializer = other.Serializer;
+        UnknownKeys = other.UnknownKeys;
+
+        SessionID = other.SessionID;
+        IsReadOnly = other.IsReadOnly;
+        IsNewSession = other.IsNewSession;
+        IsAbandoned = other.IsAbandoned;
+        Timeout = other.Timeout;
+    }
+
+    public static SessionStateCollection CreateTracking(ISessionKeySerializer serializer)
+        => new SessionStateChangeset(serializer);
+
+    public SessionStateCollection WithTracking() => new SessionStateChangeset(this);
+
+    public ISessionKeySerializer Serializer { get; }
+
+    public void AddUnknownKey(string key)
+    {
+        (UnknownKeys ??= new()).Add(key);
+    }
+
+    public void MarkUnchanged(string key) => _items[key] = new([]);
+
+    public void MarkRemoved(string key) => _items[key] = new((object?)null);
```

Comment:
Won't this cause the item to appear as new since it's using ItemHolder's object ctor?

Reply:
Yeah. It's not a real problem because of how that property is currently used, but isn't acurate. Adding some tests for that class before committing to verify behavior